src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -604,7 +619,94 @@ void QgsSymbolLegendNode::updateLabel()
+    return QString();
+
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  if ( vl && label.isEmpty() )
+  {
+    mLabel = getCurrentLabel();
+    if ( ! mLayerNode->expression().isEmpty() )
+      mLabel = evaluateLabelExpression( "[%" + mLayerNode->expression() + "%]", vl, context );
+    else if ( mLabel.contains( "[%" ) )
+      mLabel = evaluateLabelExpression( mLabel, vl, context );
+
+    emit dataChanged();
+    return mLabel;
+  }
+  else if ( vl )
+  {
+    if ( ! mLayerNode->expression().isEmpty() )
+      label = evaluateLabelExpression( label + "[%" + mLayerNode->expression() + "%]", vl, context );
+    else if ( label.contains( "[%" ) )
+      label = evaluateLabelExpression( label, vl, context );
+    return label;
+  }
+  else
+  {
+    if ( label.isEmpty() )
+      return mLabel;
+    else
+      return label;
+  }
+}
+
+QgsExpressionContext QgsSymbolLegendNode::createExpressionContext( QgsExpressionContext context ) const
+{
+
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  context.appendScope( vl->createExpressionContextScope() );
+
+  QgsExpressionContextScope *scope = new QgsExpressionContextScope( tr( "Symbol scope" ) );
+
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_label" ), getCurrentLabel().remove( "[%" ).remove( "%]" ), true ) );
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_id" ), mItem.ruleKey(), true ) );
+  QVariantList featureIds;
+  if ( vl )
+  {
+    QgsVectorLayerFeatureCounter *counter = vl->countSymbolFeatures();
+    scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_count" ), QVariant::fromValue( counter->featureCount( mItem.ruleKey() ) ), true ) );
 
+    if ( vl->featuresCounted() )
+    {
+
+      scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_feature_ids" ), featureIds, true ) );
+
+      const QgsFeatureIds fids = counter->featureIds( mItem.ruleKey() );
+
+      featureIds.reserve( fids.count() );
+      for ( QgsFeatureId fid : fids )
+      {
+        featureIds << static_cast<qint64>( fid );
+      }
+      scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_feature_ids" ), featureIds, true ) );
+    }
+    else
+    {
+      scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_feature_ids" ), featureIds, true ) );
+    }
+  }
+  else
+  {
+    scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_count" ), QVariant::fromValue( -1 ), true ) );
+    scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_feature_ids" ), featureIds, true ) );
+  }
+
+  context.appendScope( scope );
+
+  return context;
+}
+
+QString QgsSymbolLegendNode::evaluateLabelExpression( QString label, QgsVectorLayer *vl, QgsExpressionContext context ) const
+{
```

Comment:
Handles label evaluation and ensure that the proper context is generated given the type of layer.